python/ray/train/_internal/dl_predictor.py

```diff
@@ -72,14 +75,30 @@ def _predict_pandas(
         model_input = self._arrays_to_tensors(tensors, dtype)
 
         output = self.call_model(model_input)
-
         # Handle model multi-output. For example if model outputs 2 images.
-        if isinstance(output, dict):
-            return pd.DataFrame(
-                {k: TensorArray(self._tensor_to_array(v)) for k, v in output.items()}
-            )
-        else:
-            return pd.DataFrame(
-                {"predictions": TensorArray(self._tensor_to_array(output))},
-                columns=["predictions"],
-            )
+        try:
+            if isinstance(output, dict):
+                return pd.DataFrame(
+                    {
+                        k: TensorArray(self._tensor_to_array(v))
+                        for k, v in output.items()
+                    }
+                )
+            else:
+                return pd.DataFrame(
+                    {"predictions": TensorArray(self._tensor_to_array(output))},
+                    columns=["predictions"],
+                )
+        except ValueError:
+            # Fall back to returning raw pandas dataframe with user UDF output
+            if log_once("call_model_fallback_python_object"):
+                warnings.warn(
+                    "Failed to automatically cast call_model's output to a "
+                    "supported data type. Falling back to returning the raw "
+                    "python object by setting enable_tensor_extension_casting "
+                    "to False."
+                )
+            self._cast_tensor_columns = False
```

Comment:
there's probably better way like a context manager to set the global context back. I don't have much context here, open for suggestions.